src/handlers/proposalsHandlers.ts

```diff
@@ -56,7 +60,160 @@ const getProposals = (
     });
 };
 
-const getProposal = (
+export const getValueWithFullFieldFromRow = (row: ProposalRowWithFieldsAndValues):
+ProposalFieldValue => {
+  if (row.proposalVersionId === undefined
+    || row.proposalVersionId === null
+    || row.proposalVersionApplicationFormId === undefined
+    || row.proposalVersionApplicationFormId === null
+    || row.proposalVersionVersion === undefined
+    || row.proposalVersionVersion === null
+    || row.proposalVersionCreatedAt === undefined
+    || row.proposalVersionCreatedAt === null
+    || row.proposalFieldValueId === undefined
+    || row.proposalFieldValueId === null
+    || row.proposalFieldValueApplicationFormFieldId === undefined
+    || row.proposalFieldValueApplicationFormFieldId === null
+    || row.proposalFieldValueValue === undefined
+    || row.proposalFieldValueValue === null
+    || row.proposalFieldValuePosition === undefined
+    || row.proposalFieldValuePosition === null
+    || row.proposalFieldValueCreatedAt === undefined
+    || row.proposalFieldValueCreatedAt === null
+    || row.applicationFormFieldCanonicalFieldId === undefined
+    || row.applicationFormFieldCanonicalFieldId === null
+    || row.applicationFormFieldPosition === undefined
+    || row.applicationFormFieldPosition === null
+    || row.applicationFormFieldLabel === undefined
+    || row.applicationFormFieldLabel === null
+    || row.applicationFormFieldCreatedAt === undefined
+    || row.applicationFormFieldCreatedAt === null) {
+    throw new Error('Expected ProposalVersion, ProposalFieldValue, and ApplicationFormField values to be present.');
+  }
+  return {
+    id: row.proposalFieldValueId,
+    proposalVersionId: row.proposalVersionId,
+    applicationFormFieldId: row.proposalFieldValueApplicationFormFieldId,
+    position: row.proposalFieldValuePosition,
+    value: row.proposalFieldValueValue,
+    createdAt: row.proposalFieldValueCreatedAt,
+    applicationFormField: {
+      id: row.proposalFieldValueApplicationFormFieldId,
+      applicationFormId: row.proposalVersionApplicationFormId,
+      canonicalFieldId: row.applicationFormFieldCanonicalFieldId,
+      position: row.applicationFormFieldPosition,
+      label: row.applicationFormFieldLabel,
+      createdAt: row.applicationFormFieldCreatedAt,
+    },
+  };
+};
+
+const getProposalWithFieldsAndValues = (
```

Comment:
I have a not terribly helpful comment about this method, which is: I spent quite a lot of time reading through it and still couldn't figure out what it was so complicated. It seems like it should be more concise to simply collect all the versions into an array, and yet…

I trust that there is a reason for this verbosity and some of the gymnastics (like the dummy `currentProposalVersion` that gets replaced as needed). I think I would also you to do these things:

1. Consider if this method of collecting proposal versions can be simplified / made more legible.
2. If not, and/or if any improvements still aren't self-documenting, add more comments about why we're doing things.
3. Consider if anything can be abstracted out to named utility methods — even if not to reduce redundancy, just to have a clearer and more readable flow.

I leave it up to you to decide if that's worthwhile, but the long-term concern here is that this method may be difficult to maintain / mask errors.

Reply:
It is a helpful comment, thank you! One thing that comes to mind is the optional nature of the deeper objects and the mandatory (non-null) requirements of some of our types. I think the combination of those might explain some of it. But I am also unsatisfied with the way this one turned out and wish for a better way. One utility method is present, namely `getValueWithFullFieldFromRow` but there is still plenty to be desired here.